# Parallel Processing in Python

In [1]:
import random
import pandas as pd
import numpy as np
from multiprocessing import Pool

import pandas as pd

file_name = "https://github.com/har5hdeep5harma/RealWorld-DSAI-Ledger/raw/refs/heads/main/Data,%20Image%20and%20Reference%20Files/new_york_hotels.csv"
df = pd.read_csv(file_name, encoding = "ISO-8859-1")
df.head()

ean_hotel_id                                     name  \
0        269955       Hilton Garden Inn Albany/SUNY Area   
1        113431     Courtyard by Marriott Albany Thruway   
2        108151                    Radisson Hotel Albany   
3        254756  Hilton Garden Inn Albany Medical Center   
4        198232  CrestHill Suites SUNY University Albany   

                 address1    city state_province postal_code  latitude  \
0     1389 Washington Ave  Albany             NY       12206  42.68751   
1  1455 Washington Avenue  Albany             NY       12206  42.68971   
2             205 Wolf Rd  Albany             NY       12205  42.72410   
3     62 New Scotland Ave  Albany             NY       12208  42.65157   
4  1415 Washington Avenue  Albany             NY       12206  42.68873   

   longitude  star_rating  high_rate  low_rate  
0  -73.81643          3.0   154.0272  124.0216  
1  -73.82021          3.0   179.0100  134.0000  
2  -73.79822          3.0   134.1700   84.1600  
3  -73.77638          3.0   308.2807  228.4597  
4  -73.81854          3.0   169.3900   89.3900

In [2]:
# Define a basic Haversine distance formula
def haversine(df):
    lat1, lon1, lat2, lon2 = 40.671, -73.985, df['latitude'].values, df['longitude'].values
    MILES = 3959
    lat1, lon1, lat2, lon2 = map(np.deg2rad, [lat1, lon1, lat2, lon2])
    dlat = lat2 - lat1 
    dlon = lon2 - lon1 
    a = np.sin(dlat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2)**2
    c = 2 * np.arcsin(np.sqrt(a)) 
    total_miles = MILES * c
    df["distance"] = total_miles
    return df

In [3]:
%%timeit

# Vectorized implementation of Haversine applied on NumPy arrays (note we remove the values to access the numpy series)
distance_df = haversine(df)

349 μs ± 33.1 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


The slowest run took 8.71 times longer than the fastest. This could mean that an intermediate result is being cached.
1000 loops, best of 3: 349 µs per loop